### PR TITLE
Cast left node to text on string comparison

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description

Compiling the following expression fails to compile for postgres:

```sql
50 ILIKE '%5'
```

the internal expression engine will implicitly cast the left hand node to text, so we'll let postgres do the same to avoid failing and having to fallback to local evaluation.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice fix!

Does this apply to equality checks too?

I think we're missing some unit checks here, because there's a check to test if the compiled status changes for all existing expression compiler tests. So the fact that this pr passed without failing indicates that there's nothing checking this situation yet...

**Comment by m-kuhn** [comment]
Tests are on the todo list.
Can you give me one for the "equality" check and I'll see if it's also affected.

**Comment by nyalldawson** [comment]
Ok so at the moment we've only got one type conversion check: https://github.com/qgis/QGIS/blob/master/tests/src/python/featuresourcetestbase.py#L224 , and it's not impacted by this PR .

A simple equality check could be `num_char = 2`. That should not compile on postgres provider at the moment.

File: src/core/qgsfeatureiterator.h [hunk 1]
```diff
@@ -83,6 +83,14 @@ class CORE_EXPORT QgsAbstractFeatureIterator
       return mValid;
     }
 
+    /**
+     * Indicator if there was an error when sending the compiled query to the server.
+     * This indicates that there is something wrong with the expression compiler.
+     *
+     * \since QGIS 3.2
+     */
+    bool compileFailed() const;
```
Comment (nyalldawson) [inline src/core/qgsfeatureiterator.h]: We've already got a method for this - see compileStatus

File: tests/src/python/providertestbase.py [hunk 2]
```diff
@@ -817,3 +816,24 @@ def testMinMaxAfterChanges(self):
...
+                '15 NOT ILIKE \'5\'',
+                '5 ~ \'5\''):
+            iterator = self.source.getFeatures(QgsFeatureRequest().setFilterExpression('5 LIKE \'5\''))
+            count = len([f for f in iterator])
+            self.assertEqual(count, 5)
+            self.assertFalse(iterator.compileFailed())
+            if self.enableCompiler():
+                iterator = self.source.getFeatures(QgsFeatureRequest().setFilterExpression('5 LIKE \'5\''))
+                self.assertEqual(count, 5)
+                self.assertFalse(iterator.compileFailed())
```
Comment (nyalldawson) [inline tests/src/python/providertestbase.py]: Better check if the expression isn't in uncompiledFilters here - then we can indicate individual providers which don't handle this, add still catch regressions

**Comment by nyalldawson** [comment]
By the way, what do you think about removing the option to disable compilation? I think it's been default behaviour for long enough now that I'm confident there's no issues. And now so much code in QGIS depends on it, and assumes that simple field=value expressions will be fast to execute.

File: src/core/qgsfeatureiterator.h (hunk 1 above)
Comment (m-kuhn) [reply src/core/qgsfeatureiterator.h]: I know. The difference is this one reflects errors and this one is (at least on some providers) only set after starting to iterate. Would you merge the two?

File: tests/src/python/providertestbase.py (hunk 2 above)
Comment (m-kuhn) [reply tests/src/python/providertestbase.py]: Given travis is happy, is this a requirement?

**Comment by m-kuhn** [comment]
Removing the option would be ok for me.

File: tests/src/python/providertestbase.py (hunk 2 above)
Comment (nyalldawson) [reply tests/src/python/providertestbase.py]: Guess not... I'm just surprised no other providers have issues with this. I'd have thought some would. Maybe mssql/oracle which aren't tested on Travis (But who cares about those :P )

Comment (m-kuhn) [reply tests/src/python/providertestbase.py]: Well, even if they suffer from this, the result will be correct (just slower). The test will only trigger once they start returning `true` for `compileFailed()`, right now they will silently fall back to local filtering.

**Pull request merged**